### ISSUE TITLE
Slow queue wait implementation causes delay in processing of new main queue messages

### ISSUE THREAD

**Issue by dougal** [iHiD/propono#97]
If the main queue is empty after 20 seconds, the client then listens on the slow queue for 20 seconds. While listening on the slow queue, any item arriving in the main queue will have to wait till the slow queue listen is over, and the next main queue listen, to be processed.

Timeline:

1. Listener starts to wait on main queue
2. Nothing to process
3. Listener stops listening on main queue after 20 seconds
4. Listener starts to wait on slow queue
5. Nothing to process
6. After 1 second an item is added to the main queue
7. Listener stops listening on the slow queue after a further 19 seconds
8. Listener starts to wait on main queue
9. Item in main queue is immediately received and processed, 19 seconds after it was published.

Code to show in action:

```
topic  = :slow_queue_delay_test
client = Propono::Client.new

Thread.new do
  client.listen(topic) do |payload|
    puts "Message processed after #{Time.now.to_i - payload[:time]} seconds"
  end
end

# One second longer than the long-poll time.
sleep 21
PublishProponoMessage.(topic, { time: Time.now.to_i })

# Keep the main thread alive.
sleep 30
```

Suggested Solution:

Give the main queue a 20 second head-start, before listening on both queues. Reset this head-start after ever job processed.

1. Listen to the main queue for 20 seconds.
2. Listen to the main AND slow queues indefinitely.
3. If a job turns up in one queue or the other, process it, go back to (1).

**Comment by iHiD** [comment]
I'll admit that my solution is marginally more brute force to yours: https://github.com/iHiD/propono/commit/32690a6c5f7f118523731a8dbcaef825af63311f :)

**Comment by dougal** [comment]
Closed by 32690a6.
